File: C++/Caesar.cc
#include <iostream>
#include <string>
using namespace std;

char alpha [26] = {'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 
                   'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 
                   'w', 'x', 'y', 'z'};

int mod(int n, int m)
{
    return ((n % m) + m) % m;
}

class Caesar
{
    private:
        int shift;
        char *alphabet;
    public:
        Caesar(int);
        Caesar(int, char*); 
        string encrypt(string);
        string decrypt(string);
        int indexOf(char); //should be changed to T type
};

Caesar::Caesar(int shiftVal)
{
    shift = shiftVal;
    alphabet = alpha;
}

Caesar::Caesar(int shiftVal, char *alphaVal)
{
    shift = shiftVal;
    alphabet = alphaVal;
}

int Caesar::indexOf(char character)
{
    for (int i = 0; i < 26; ++i)
    {
        if (character == alphabet[i]) // change the alphabet order according to frequency
        {
            return i;
        } 
    }
    
    return -1;
}

string Caesar::encrypt(string plaintext)
{
    string ciphertext ="";
    int i, index;

    for (i=0; i<plaintext.size(); ++i)
    {
        index = Caesar::indexOf(plaintext[i]);
        index = mod((index + shift), 26);
        ciphertext = ciphertext + alphabet[index];
    }
    return ciphertext;
}

string Caesar::decrypt(string ciphertext)
{
    string plaintext ="";
    int i, index;

    for (i=0; i<ciphertext.size(); ++i)
    {
        index = Caesar::indexOf(ciphertext[i]);
        index = mod((index - shift), 26);
        plaintext = plaintext + alphabet[index];
    }
    return plaintext;
}

class Exploits
{
    private:
        char *alphabet;
    public:
        Exploits();
        Exploits(char*);
        int indexOf(char);
        void ciphertext_only(string);
        int known_plaintext(string, string);
        int chosen_plaintext(Caesar);
        int chosen_ciphertext(Caesar);
};


int Exploits::indexOf(char character)
{
    for (int i = 0; i < 26; ++i)
    {
        if (character == alphabet[i]) // change the alphabet order according to frequency
        {
            return i;
        } 
    }
    
    return -1;
}

Exploits::Exploits()
{
    alphabet = alpha;
}

Exploits::Exploits(char *alphaVal)
{
    alphabet = alphaVal;
}

void Exploits::ciphertext_only(string ciphertext)
{
    int i;
    for(i=0; i < 26; ++i)
    {
        Caesar c (i);
        cout << c.decrypt(ciphertext) + "\n";
    }
}

int Exploits::known_plaintext(string plaintext, string ciphertext)
{
    int i = mod((indexOf(ciphertext[0]) - indexOf(plaintext[0])), 26); 
    return i;
}

int Exploits::chosen_plaintext(Caesar cipher)
{
    string a (1,alphabet[0]);
    string c = cipher.encrypt(a);
    int i = indexOf(c[0]);
    return i;
}

int Exploits::chosen_ciphertext(Caesar cipher)
{
    string a (1,alphabet[0]);
    string c = cipher.decrypt(a);
    int i = indexOf(c[0]);
    i = mod(i * (-1), 26);
    return i;
}

int main() 
{
    Caesar c (3);
    cout << c.encrypt("hello") + "\n";
    cout << c.decrypt("khoor") + "\n";
    Exploits e;
    e.ciphertext_only("khoor");
    cout << e.known_plaintext("hello", "khoor");
    cout << "\n";
    cout << e.chosen_plaintext(c);
    cout << "\n";
    cout << e.chosen_ciphertext(c);
    cout << "\n";
}

File: Ruby/Caesar.rb

def mod(n, m)
    return ((n % m) + m) % m
end

class Caesar

    def initialize(shiftVal, alphaVal=('a'..'z').to_a)
        @shift = shiftVal
        @alphabet = alphaVal
    end
    
    def encrypt(plaintext)
        ciphertext = ""
        for i in 0...plaintext.length
            key = @alphabet.index(plaintext[i]);
            key = mod(key+@shift, @alphabet.length)
            ciphertext = ciphertext + @alphabet[key]
        end
        return ciphertext
    end

    def decrypt(ciphertext)
        plaintext = ""
        for i in 0...ciphertext.length
            key = @alphabet.index(ciphertext[i]);
            key = mod(key-@shift, @alphabet.length)
            plaintext = plaintext + @alphabet[key]
        end
        return plaintext
    end
end

class Exploits

    def initialize(alphaVal=('a'..'z').to_a)
        @alphabet = alphaVal
    end

    def ciphertext_only(ciphertext)
        for i in 0...@alphabet.length
            c = Caesar.new(i)
            print c.decrypt(ciphertext) + "\n"
        end 
    end
    
    def known_plaintext(plaintext, ciphertext)
        i = @alphabet.index(plaintext[0])
        j = @alphabet.index(ciphertext[0])
        key = mod(j-i, @alphabet.length)
        return key
    end

    def chosen_plaintext(cipher)
        letter = cipher.encrypt(@alphabet[0])
        return @alphabet.index(letter)
    end
    
    def chosen_ciphertext(cipher)
        letter = cipher.decrypt(@alphabet[0])
        index =  @alphabet.index(letter)
        index = mod(index * (-1), @alphabet.length)
        return index
    end

end

c = Caesar.new(3)
print c.encrypt("hello")
print "\n"
print c.decrypt("khoor")
print "\n"
e = Exploits.new()
e.ciphertext_only("khoor")
print e.known_plaintext("hello","khoor")
print "\n"
print e.chosen_plaintext(c)
print "\n"
print e.chosen_ciphertext(c)
print "\n"
